changefeedccl: Async flushes may produce out of order events

Failure observed on master:
```
Failed
=== RUN   TestChangefeedNemeses/cloudstorage
    helpers_test.go:743: making server as system tenant
    helpers_test.go:830: making cloudstorage feed factory
    nemeses_test.go:40: topic foo partition : saw new row timestamp 1665420895637060333.0000000000 after 1665420896550554241.0000000000 was seen
    nemeses_test.go:40: topic foo partition : saw new row timestamp 1665420895561568242.0000000000 after 1665420896550554241.0000000000 was seen
    --- FAIL: TestChangefeedNemeses/cloudstorage (11.99s)
```

The test ran with parallel workers, and async flushing enabled.

My best guess is that async flushing could produce the following:
  k@t1 is written to sink and causes flush f1
  k@t2 is written to sink and causes flush f2
  f2 completes before f1

(thanks @ajwerner for reporting this)

  

Jira issue: CRDB-20371

--- a/pkg/ccl/changefeedccl/sink_cloudstorage.go
+++ b/pkg/ccl/changefeedccl/sink_cloudstorage.go
@@ -18,7 +18,6 @@ import (
 	"net/url"
 	"path/filepath"
 	"strings"
-	"sync"
 	"sync/atomic"
 	"time"
 
@@ -31,6 +30,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/humanizeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -307,9 +307,16 @@ type cloudStorageSink struct {
 	metrics           metricsRecorder
 
 	asyncFlushActive bool
-	flushCtx         context.Context
-	flushGroup       sync.WaitGroup
-	flushErr         atomic.Value
+	flushGroup       ctxgroup.Group
+	asyncFlushCh     chan flushRequest // channel for submitting flush requests.
+	asyncFlushTermCh chan struct{}     // channel closed by async flusher to indicate an error
+	asyncFlushErr    error             // set by async flusher, prior to closing asyncFlushTermCh
+}
+
+type flushRequest struct {
+	file  *cloudStorageSinkFile
+	dest  string
+	flush chan struct{}
 }
 
 var cloudStorageSinkIDAtomic int64
@@ -324,6 +331,16 @@ var partitionDateFormats = map[string]string{
 	"hourly": "2006-01-02/15/",
 }
 var defaultPartitionFormat = partitionDateFormats["daily"]
+
+// flushQueueDepth puts a limit on how many flush requests
+// may be outstanding, before we block.
+// In reality, we will block much sooner than this limit due
+// to blocking buffer memory limits (in its default configuration);
+// We just want this setting to be sufficiently large, but not
+// so large as to have extremely large flush queues.
+// The default of 256, with the default file size of 16MB, gives us
+// a queue of 2.5GB of outstanding flush data.
+const flushQueueDepth = 256
 
 func makeCloudStorageSink(
 	ctx context.Context,
@@ -372,8 +389,11 @@ func makeCloudStorageSink(
 		topicNamer:       tn,
 		asyncFlushActive: enableAsyncFlush.Get(&settings.SV),
 		// TODO (yevgeniy): Consider adding ctx to Dial method instead.
-		flushCtx: ctx,
+		flushGroup:       ctxgroup.WithContext(ctx),
+		asyncFlushCh:     make(chan flushRequest, flushQueueDepth),
+		asyncFlushTermCh: make(chan struct{}),
 	}
+	s.flushGroup.GoCtx(s.asyncFlusher)
 
 	if partitionFormat := u.consumeParam(changefeedbase.SinkParamPartitionFormat); partitionFormat != "" {
 		dateFormat, ok := partitionDateFormats[partitionFormat]
@@ -385,8 +405,7 @@ func makeCloudStorageSink(
 	}
 
 	if s.timestampOracle != nil {
-		s.dataFileTs = cloudStorageFormatTime(s.timestampOracle.inclusiveLowerBoundTS())
-		s.dataFilePartition = s.timestampOracle.inclusiveLowerBoundTS().GoTime().Format(s.partitionFormat)
+		s.setDataFileTimestamp()
 	}
 
 	switch encodingOpts.Format {
@@ -520,7 +539,7 @@ func (s *cloudStorageSink) EmitResolvedTimestamp(
 
 	// Wait for previously issued async flush requests to complete
 	// before we write  resolved time stamp file.
-	if err := s.waitAsyncFlush(); err != nil {
+	if err := s.waitAsyncFlush(ctx); err != nil {
 		return errors.Wrapf(err, "while emitting resolved timestamp")
 	}
 
@@ -587,13 +606,17 @@ func (s *cloudStorageSink) Flush(ctx context.Context) error {
 		return err
 	}
 	s.files.Clear(true /* addNodesToFreeList */)
+	s.setDataFileTimestamp()
+	return s.waitAsyncFlush(ctx)
+}
 
+func (s *cloudStorageSink) setDataFileTimestamp() {
 	// Record the least resolved timestamp being tracked in the frontier as of this point,
 	// to use for naming files until the next `Flush()`. See comment on cloudStorageSink
 	// for an overview of the naming convention and proof of correctness.
-	s.dataFileTs = cloudStorageFormatTime(s.timestampOracle.inclusiveLowerBoundTS())
-	s.dataFilePartition = s.timestampOracle.inclusiveLowerBoundTS().GoTime().Format(s.partitionFormat)
-	return s.waitAsyncFlush()
+	ts := s.timestampOracle.inclusiveLowerBoundTS()
+	s.dataFileTs = cloudStorageFormatTime(ts)
+	s.dataFilePartition = ts.GoTime().Format(s.partitionFormat)
 }
 
 // enableAsyncFlush controls async flushing behavior for this sink.
@@ -605,13 +628,27 @@ var enableAsyncFlush = settings.RegisterBoolSetting(
 )
 
 // waitAsyncFlush waits until all async flushes complete.
-func (s *cloudStorageSink) waitAsyncFlush() error {
-	s.flushGroup.Wait()
-	if v := s.flushErr.Load(); v != nil {
-		return v.(error)
+func (s *cloudStorageSink) waitAsyncFlush(ctx context.Context) error {
+	done := make(chan struct{})
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-s.asyncFlushTermCh:
+		return s.asyncFlushErr
+	case s.asyncFlushCh <- flushRequest{flush: done}:
 	}
-	return nil
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-s.asyncFlushTermCh:
+		return s.asyncFlushErr
+	case <-done:
+		return nil
+	}
 }
+
+var logQueueDepth = log.Every(30 * time.Second)
 
 // flushFile flushes file to the cloud storage.
 // file should not be used after flushing.
@@ -620,7 +657,7 @@ func (s *cloudStorageSink) flushFile(ctx context.Context, file *cloudStorageSink
 	if s.asyncFlushActive && !asyncFlushEnabled {
 		// Async flush behavior was turned off --  drain any active flush requests
 		// before flushing this file.
-		if err := s.waitAsyncFlush(); err != nil {
+		if err := s.waitAsyncFlush(ctx); err != nil {
 			return err
 		}
 	}
@@ -648,21 +685,63 @@ func (s *cloudStorageSink) flushFile(ctx context.Context, file *cloudStorageSink
 		return file.flushToStorage(ctx, s.es, dest, s.metrics)
 	}
 
-	s.flushGroup.Add(1)
-	go func() {
-		defer s.flushGroup.Done()
-		// NB: must use s.flushCtx; ctx may be short lived (i.e. cancelled).
-		if err := file.flushToStorage(s.flushCtx, s.es, dest, s.metrics); err != nil {
-			log.Errorf(ctx, "error flushing file to storage: %s", err)
-			// We must use the same type for error we store in flushErr.
-			s.flushErr.CompareAndSwap(nil, &flushError{error: err})
+	// Try to submit flush request, but produce warning message
+	// if we can't.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-s.asyncFlushTermCh:
+		return s.asyncFlushErr
+	case s.asyncFlushCh <- flushRequest{file: file, dest: dest}:
+		return nil
+	default:
+		if logQueueDepth.ShouldLog() {
+			log.Infof(ctx, "changefeed flush queue is full; ~%d bytes to flush",
+				flushQueueDepth*s.targetMaxFileSize)
 		}
-	}()
-	return nil
+	}
+
+	// Queue was full, block until it's not.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-s.asyncFlushTermCh:
+		return s.asyncFlushErr
+	case s.asyncFlushCh <- flushRequest{file: file, dest: dest}:
+		return nil
+	}
 }
 
-type flushError struct {
-	error
+func (s *cloudStorageSink) asyncFlusher(ctx context.Context) error {
+	defer close(s.asyncFlushTermCh)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case req, ok := <-s.asyncFlushCh:
+			if !ok {
+				return nil // we're done
+			}
+
+			// handle flush request.
+			if req.flush != nil {
+				close(req.flush)
+				continue
+			}
+
+			// flush file to storage.
+			flushDone := s.metrics.recordFlushRequestCallback()
+			err := req.file.flushToStorage(ctx, s.es, req.dest, s.metrics)
+			flushDone()
+
+			if err != nil {
+				log.Errorf(ctx, "error flushing file to storage: %s", err)
+				s.asyncFlushErr = err
+				return err
+			}
+		}
+	}
 }
 
 // flushToStorage writes out file into external storage into 'dest'.
@@ -695,7 +774,10 @@ func (f *cloudStorageSinkFile) flushToStorage(
 // Close implements the Sink interface.
 func (s *cloudStorageSink) Close() error {
 	s.files = nil
-	return errors.CombineErrors(s.waitAsyncFlush(), s.es.Close())
+	err := s.waitAsyncFlush(context.Background())
+	close(s.asyncFlushCh) // signal flusher to exit.
+	err = errors.CombineErrors(err, s.flushGroup.Wait())
+	return errors.CombineErrors(err, s.es.Close())
 }
 
 // Dial implements the Sink interface.

--- a/pkg/ccl/changefeedccl/sink_cloudstorage_test.go
+++ b/pkg/ccl/changefeedccl/sink_cloudstorage_test.go
@@ -197,7 +197,9 @@ func TestCloudStorageSink(t *testing.T) {
 	testWithAndWithoutAsyncFlushing := func(t *testing.T, name string, testFn func(*testing.T)) {
 		t.Helper()
 		testutils.RunTrueAndFalse(t, name+"/asyncFlush", func(t *testing.T, enable bool) {
+			old := enableAsyncFlush.Get(&settings.SV)
 			enableAsyncFlush.Override(context.Background(), &settings.SV, enable)
+			defer enableAsyncFlush.Override(context.Background(), &settings.SV, old)
 			testFn(t)
 		})
 	}
@@ -460,7 +462,7 @@ func TestCloudStorageSink(t *testing.T) {
 	})
 
 	waitAsyncFlush := func(s Sink) error {
-		return s.(*cloudStorageSink).waitAsyncFlush()
+		return s.(*cloudStorageSink).waitAsyncFlush(context.Background())
 	}
 	testWithAndWithoutAsyncFlushing(t, `bucketing`, func(t *testing.T) {
 		t1 := makeTopic(`t1`)

